tests/test_bosonic_solver.py

```diff
@@ -3,23 +3,157 @@
 """
 import numpy as np
 from numpy.linalg import eigvalsh
+from scipy.integrate import quad
+from qutip import (Qobj, QobjEvo, sigmaz, sigmax, basis, expect, Options, destroy, basis)
+from bofin.heom import add_at_idx
+from bofin.heom import HSolverDL
+from bofin.heom import BosonicHEOMSolver
+from bofin.heom import _heom_state_dictionaries, _check_Hsys
+import pytest
+from math import sqrt, factorial
 
+def test_add_at_idx():
+    """
+    Tests the function to add at hierarchy index.
+    """
+    seq = (2, 3, 4)
+    assert add_at_idx(seq, 2, 1), (2, 3, 5)
 
-from scipy.integrate import quad
+    seq = (2, 3, 4)
+    assert add_at_idx(seq, 0, -1), (1, 3, 4)
 
+def test_state_dictionaries():
+    """
+    Tests the _heom_state_dictionaries.
+    """
 
-from qutip import Qobj, sigmaz, sigmax, basis, expect, Options, destroy
 
+    #bosonic
+    kcut = 6
+    N_cut = 4
+    nhe, he2idx, idx2he = _heom_state_dictionaries(
+            [N_cut + 1] * kcut, N_cut
+        )
+        
+    total_nhe = int(
+            factorial(N_cut + kcut)
+            / (factorial(N_cut) * factorial(kcut))
+        )
+    assert nhe, total_nhe
+        
+def test_check_H():
+    """Tests the function for checking system Hamiltonian"""
+    assert(_check_Hsys(sigmax()))
+    assert(_check_Hsys([sigmax(), sigmaz()]))
+    assert(_check_Hsys([[sigmax(), np.sin], [sigmaz(), np.cos]]))
+    assert(_check_Hsys([[sigmax(), np.sin], [sigmaz(), np.cos]]))
+    assert(_check_Hsys(QobjEvo([sigmaz(), sigmax(), sigmaz()])))
```

Comment:
```suggestion
    _check_Hsys(sigmax())
    _check_Hsys([sigmax(), sigmaz()])
    _check_Hsys([[sigmax(), np.sin], [sigmaz(), np.cos]])
    _check_Hsys([[sigmax(), np.sin], [sigmaz(), np.cos]])
    _check_Hsys(QobjEvo([sigmaz(), sigmax(), sigmaz()]))
```
The `_check_Hsys` function should either return `None` and raise an exception if the check fails, or return `True` or `False`. Currently it returns `True` and raises an exception which mixes the two patterns in a way that makes the returned value redundant -- if `_check_Hsys` returns at all the check has already passed and the return value is irrelevant.

Reply:
yeah, this sounds sensible!